### PR TITLE
SAM-64: Changes to support UI and JSON Schema with versioned dist via GHA

### PR DESCRIPTION
**Pull request by eapearson**
# Summary of Changes

## TLDR

This set of changes is oriented around adding support for the exporting of sample services specs into a less idiosyncratic, more ui-friendly sample configuration files. The files will be made available by new sample service methods which essentially consume and then emit those files.

The existing generated files and their location are retained for usage by the Sample Service and Sample Uploader.

This PR implements this through a new set of python and shell scripts, and a configuration distribution process based on GitHub Action workflows which create distrubution branches under various conditions.

The new config files support field ordering and grouping (`groups.json`), and field definitions (`schema.json`), into json files more suitable for direct ui consumption. The grouping is a simple json object. The field definitions are jsonschema with KBase enhancements.

The jsonshchema may be used validate samples, although no production code does or is planned to do so. However, the process of developing the jsonschema, as well as jsonschema to validate the jsonschema (!) revealed issues in the sample configuration that needed correction:

- Added missing fields
- Removed duplicated fields
- Ignore SESAR fields which aren't part of the user sample
- Make some attribute names more jsonchema-friendly
- Remove aliases which are duplicative (most of them)
- For SESAR, add an apiKey to capture the field names for bulk sample download and import

A few miscellaneous code quality and other fixes.

I've tested these changes, or at least the resulting output files, with a sample validation and import tool utilizing 10s of thousands of SESAR samples. At present these samples fully validate, with the exception of a handful of samples with true validation errors. They do not presently import into our systems due to the fact that some "unit" fields have been restored and are currently (outside of this PR) not configured to be present in samples. Thus, when placed in a sample's meta_controlled the sample service throws an error:

```
Cannot validate controlled field "sesar:elevation_unit", no matching validator found
```

### Danger

Some of these changes may be breaking:

- sample template column label matching is assumed to be case insensitive; there were many field name aliases which differed just by case - any such duplicates were removed (it is an endless rabbit hole to provide unique case-convention versions of each such alias.)

## templates

- For both templates, moved the top level `Template`, `ID`, and `Name` fields to the top of the file. It is simply easier to grasp the template with these fields at the top of the file, and the columns extending below them.

- For both templates, added the top level `Parent` field. We are not currently using this feature, but we do have existing mappings for this, so I've restored them. If we fail to capture this info those samples will never be able to form relationships.

- For both templates, field labels are considered to be compared case insensitively; any alias which duplicates the primary field label, or another alias, is removed; any empty aliases attribute is removed.

### ENIGMA

No other changes

### SESAR

- add `apiKey` mandatory attribute; under certain conditions (bulk download) the downloaded samples are formatted as csv with column labels in the form of the api keys, which are snake-cased. The other form of download is Excel xlsx, which uses the human-friendly form of column labels.
    - the apiKey was added by noting samples as they appear in bulk, and also by fetching samples via the api.
    - may require some additional work; e.g. not propagated to the validators that the uploader users.

- add `disposition` attribute, which contains a disposition symbol (string): `required`, `recommended`, `ignore`. 
    - `required` and `recommended` reflect the SESAR documentation.
    - `required` should be used to ensure that SESAR-required fields are present; I've found, though, that SESAR does not enforce this across all sample types. Imagine a parent sample for a site, with multiple child samples for actual specimens.
    - `recommended` could be used as a quality measurement
    - `ignore` means that the field should be ignored and not imported; see the "URL" comment in "Thoughts" below.

- restore the fields `age_unit` and `geological_unit`, which are used to provide the unit for their associated value fields; if we don't store all fields uploaded, downloading sample spreadsheets which can be used for re-import is more difficult.
    - I understand why this was done, because the unit is folded into the target value field, but I think we should keep those fields so we can display them, and if we support downloading update spreadsheets, they do need to be present.

- restore the transformation for `Comment`, not sure why it was missing.

- added missing fields:
    - URL
        - ignored, since it is a synthesized field, created from the base url for viewing a sample, with the IGSN for the given sample inserted.
        - this is the "build_url" and is not the same as the "related url" which the config previously mapped to `url` (and should also be ignored.)

#### Ignoring Fields

The new template attribute "disposition" allows ignoring fields. Here are some examples of fields which should be ignored:

- `Cur Registrant Name` set as disposition "ignore". It is generated from the user's id stored with the sample at upload time. There are a few problems with this field:
    -  We don't have a use for this field
    -  it may change over time (e.g. user changes name at SESAR.)
    -  it is not present in the initial sample upload template

- The "Related URL" fields are ignored. This includes:
    - Related URL
    - Related URL Type
    - Related URL Description
    
  These fields are supported for initial upload, but are not downloaded. They are available through the api as either "external_url" or "publicationUrl", each of which support multiple urls.
  
  They are not available for download, probably because the internal structure is actually a set of related urls and css doesn't explicitly support a sub-table in a field. The spreadsheet upload allows a single related url.
  
  It is notable that SESAR also supports uploading by XML, which does support more complex data than a flat table. E.g. it allows uploading of multiple related urls.

- `is_archived`: Appears to be another synthesized field; it is not present in the upload template.
    - nor have I encountered a sample with a value for this field, out of 10s of thousands looked at.

## Export Scripts

Scripts were refactored. There are now three sets of scripts:

- scripts/export - Python scripts to export sample specs for service, ui, and documentation purposes
- scripts/validate - Python scripts to validate templates, and re-usable scripts for validating json and yaml files against jsonschema
- scripts/automation - Shell scripts which invoke export and validation scripts, and handle other building tasks.

The automation scripts will place all built assets into the `dist` directory, which is excluded from git. For a local build, the dist directory will simply be populated. For a GHA build, the directory is populated, and then copied to a distribution branch.

The new scripts which generate and validate the ui configs are located in `scripts/export/ui` , where you can find a `README.md` describing each script.

## Existing Scripts

Existing scripts were only lightly touched, and simply relocated.

- `/scripts/validate/validate_templates.py` required adding `'Parent'` as a required field; touching the file required some code style changes (automated)

## Vocabulary changes

A few changes were made to vocabularies. This includes:
- some bug fixes (ENIGMA units)

## Field Grouping / Ordering Changes

Changes consist of adding missing fields, and in one case, a missing group. The `scripts/export/ui/export.py` script checks for the consistency of groups and controlled field schemas, ensuring that each controlled field exists in the grouping config, and that the grouping config only contains controlled fields.

## Spec Distribution

A GitHub Action workflow creates distribution branches containing just the built sample service specs. 

Please see the dedicated document docs/build-and-deploy.md for details.

### CODE REVIEW

**Comment by eapearson** [comment]
I realize the checks (tests) are failing, working on a GHA based test runner (have working Docker image config, just need to wire up the workflow.)

**Comment by eapearson** [comment]
@scanon, sorry, I thought you might be pinged on this for review, so consider this the ping!

**Comment by eapearson** [comment]
I realize that the new eap_scripts do not permanently save the results (as the other config transformation scripts do), and the directory should probably not be named eap_scripts. 
The original vision is that the sample_service could do that transformation on the fly. So some form of that code could live in the sample_service. But I think it might be better if, the sample_service, or whoever uses the config repo, after cloning the config repo, could simply run those scripts to generate the files. The generated files are the same ones used by the sample and sampleset landing pages, as well as the experimental/testing importer I've been using to exercise them against many SESAR samples.

File: vocabularies/core.yml [hunk 1]
```diff
@@ -2,13 +2,18 @@ terms:
   biome:
     description: Biome from ENVO (child of ENVO:00000428)
     examples:
-    - FENVO:01001831
-    ontology:
```
Comment (scanon) [inline vocabularies/core.yml:6]: Is there a reason for the change?  I guess it is fine but having them nested seemed nice in some ways.

File: scripts/merge_validators.py [hunk 2]
```diff
@@ -63,16 +62,17 @@ def expand_validators(val):
...
-        v['parameters'] = {
-                'ontology': val['ontology']['ns'],
-                'ancestor_term': val['ontology']['ancestor_term']
-                }
     elif 'max-len' in val:
-             v['parameters'] = {'max-len': val['max-len']}
+        v['parameters'] = {'max-len': val['max-len']}
     elif typ == 'string':
-        return deepcopy(_NOOP)
+        if format == 'ontology-term':
```
Comment (scanon) [inline scripts/merge_validators.py:68]: Hmm.  So instead of saying this is a type ontology you are making ontology sort of a subtype?  I'm not sure if I like this or not.

**Review by scanon — COMMENTED** [review]
See my comments so far.  I haven't looked closely at your scripts yet.  So we can follow up on those.

File: scripts/merge_validators.py (hunk 2 above)
Comment (eapearson) [reply scripts/merge_validators.py:68]: That is a jsonschema thing. (But I understand the impulse - when I put together a first draft of my own mini-type system before moving towards jsonschema and jsonschema compatibility I did the same thing -- an ontology term is its own type.)

In jsoncschema one cannot add new types -- the types are the physical storage type which is fixed for JSON. The format is typically used for things like url, email, timestamp, etc. (https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.7.3), which describes a "string format", that is, a constrained form of a string. 

> The "format" annotation keyword is defined to allow schema authors to convey semantic information for a fixed subset of values which are accurately described by authoritative resources, be they RFCs or other external specifications.

Formats are allowed for other types as well, but the standard does not predefine any. The standard predefines some formats (as cited above), but does not require that implementations support them.

`format` does seem applicable to our ontology terms, which is a string with a specific format, and has an external definition. The tying of other custom attributes (namespace, etc.) as required for validation of a value in the format is not typical. I did not dig deeply into the spec to determine if there are other restrictions or recommendations - since addition of custom attributes is allowed, I simply elected to move the ontology-term defining fields namespace and ancestorTerm from the "ontology" object into the top level. 

I could have left it as-is, and only done that transformation later in the export scripts, but I thought part of this effort was to move the spec format towards jsonschema compatibility, as you had already been doing work towards that end.

File: vocabularies/core.yml (hunk 1 above)
Comment (eapearson) [reply vocabularies/core.yml:6]: See the comment below.

This is just an effort to have the structure be more similar to jsonschema, which, afaik, favors constraint attributes at the top level for a type - at least that is the only style used in the standard.
One of the reasons may be that jsonschemas can become quite nested, and the lower internal nesting at the type level helps simplify the overall structure.

File: eap_scripts/README.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+Transforms template definitions found in `/templates` to `DIR/formats`.
+
+Although now obsolete from front end consumption (as formats have been removed as a first-class concept in samples), it is useful for experimental importers.
+
+## `verify.py`
+
+Requires the above transformations to have already been conducted.
+
+
+## Adding apiKey
```
[the remark this answers is not in the document]
Comment (eapearson) [reply eap_scripts/README.md:30]: There is actually a second SESAR api which uses different keys.
Upon discovering this I did not change `apiKey`, but considered that there may be use cases for multiple api keys.
The reason for a bespoke "api key" as opposed to simply "aliases" or "other keys", is that an api key has a very specific, strict, and useful meaning compared to aliases. 

An alternative structure could be:

apiKeys:
   - type: v0
      key: depth_min
  - type: v1
     key: depthMin

It gets more complicated, though, as we can expect different versions of an api to handle data differently. E.g. in the case above, the v0 (unversioned, actually) SESAR api returns depth_min as a string, but deptMin as a number.

In the end, I considered this a way to support the SESAR format specifically, in a manner more structured and easier to manage than using the aliases, and only SESAR data exported from the web interface.

Configuration support directly for apis would require more stuff anyway, and probably would require an separate schema per api.

In retrospect, as I dug further into SESAR, this would perhaps be more aptly named "bulkKey", since the keys are those emitted when samples are downloaded in bulk. I named it "apiKey" because they seem to be identical to the keys in the v0 api.

**Comment by eapearson** [comment]
@Shane  I tried to address all of your comments. 

The PR is on the big side, and the review comment UI is a bit finicky, so I'm not 100% sure I addressed all the questions or comments.

**Comment by eapearson** [comment]
The last set of changes adds GHA workflow support for building the distributable assets - the metadata validation / definition files, the groupings, the templates, and an html page presenting the sample fields in the grouping order.

While doing this work, I also reorganized files and removed unused data, script and other files. 

The files are published to a new set of branches, `dist-master`, `dist-release`, `dist-pull_request-#` (where # is the PR #). In addition the release is tagged with the version `dist-release-v#.#.#`.

Other than the new location for these files, the formats are the same. The ontology metadata file was removed since it is duplicative of the main metadata validation file and was not covered by a jsonschema (which is used for validation of file format.)

Documentation explaining this further has been added in the docs directory.
